sdk/metric/instrument.go

```diff
@@ -24,10 +24,157 @@ import (
+	// are measuring.
+	InstrumentKindSyncHistogram
+	// InstrumentKindAsyncCounter identifies a group of instruments that record
+	// increasing values in an asynchronous callback.
+	InstrumentKindAsyncCounter
+	// InstrumentKindAsyncUpDownCounter identifies a group of instruments that
+	// record increasing and decreasing values in an asynchronous callback.
+	InstrumentKindAsyncUpDownCounter
+	// InstrumentKindAsyncGauge identifies a group of instruments that record
+	// current values in an asynchronous callback.
+	InstrumentKindAsyncGauge
+)
+
+type nonComparable [0]func() // nolint: unused  // This is indeed used.
+
+// Instrument describes properties an instrument is created with.
+type Instrument struct {
+	// Name is the human-readable identifier of the instrument.
+	Name string
+	// Description describes the purpose of the instrument.
+	Description string
+	// Kind defines the functional group of the instrument.
+	Kind InstrumentKind
+	// Unit is the unit of measurement recorded by the instrument.
+	Unit unit.Unit
+	// Scope identifies the instrumentation that created the instrument.
+	Scope instrumentation.Scope
+
+	// Ensure forward compatibility if non-comparable fields need to be added.
+	nonComparable // nolint: unused
+}
+
+// mask returns a copy of base with all non-zero-value fields of m replacing
+// the fields of the returned copy.
+func (base Instrument) mask(m Instrument) Instrument {
+	cp := base
+	if m.Name != "" {
+		cp.Name = m.Name
+	}
+	if m.Description != "" {
+		cp.Description = m.Description
+	}
+	if m.Kind != zeroInstrumentKind {
+		cp.Kind = m.Kind
+	}
+	if m.Unit != zeroUnit {
+		cp.Unit = m.Unit
+	}
+	if m.Scope.Name != "" {
+		cp.Scope.Name = m.Scope.Name
+	}
+	if m.Scope.Version != "" {
+		cp.Scope.Version = m.Scope.Version
+	}
+	if m.Scope.SchemaURL != "" {
+		cp.Scope.SchemaURL = m.Scope.SchemaURL
+	}
+	return cp
+}
+
+// empty returns if all fields of i are their zero-value.
+func (i Instrument) empty() bool {
+	return i.Name == "" &&
+		i.Description == "" &&
+		i.Kind == zeroInstrumentKind &&
+		i.Unit == zeroUnit &&
+		i.Scope == zeroScope
+}
+
+// matches returns whether all the non-zero-value fields of i match the
+// corresponding fields of other. If i is empty it will match all other, and
+// true will always be returned.
+func (i Instrument) matches(other Instrument) bool {
+	return i.matchesName(other) &&
+		i.matchesDescription(other) &&
+		i.matchesKind(other) &&
+		i.matchesUnit(other) &&
+		i.matchesScope(other)
+}
+
+// matchesName returns true if the Name of i is "" or it equals the Name of
+// other, otherwise false.
+func (i Instrument) matchesName(other Instrument) bool {
+	return i.Name == "" || i.Name == other.Name
+}
+
+// matchesDescription returns true if the Description of i is "" or it equals
+// the Description of other, otherwise false.
+func (i Instrument) matchesDescription(other Instrument) bool {
+	return i.Description == "" || i.Description == other.Description
```

Comment:
Do we think anyone will want to explicitly match on empty values?

Reply:
I think that's possible, but not sure it will be likely.

I expected if that were the case a user would write their own `View`.

Reply:
Added a comment to `NewView` informing about how to create a zero-value matching view (i.e. create their own `View`).